src/antlr/antlr.test.ts

```diff
@@ -0,0 +1,137 @@
+import { CharStreams, CommonTokenStream } from "antlr4ts";
+import { UppercaseCharStream } from "./UppercaseCharStream";
+import { SqlBaseLexer } from "./SqlBaseLexer";
+import { SqlBaseParser } from "./SqlBaseParser";
+import { getLineage } from "./LineageUtil";
+import { Lineage, Table } from "../Lineage";
+
+type ColumnId = string;
+const columnsMapping: { [tableId: string]: ColumnId[] } = {
+  account: ["account_type", "account_id", "account_description", "account_parent", "account_rollup"],
+  employee: [
+    "birth_date",
+    "department_id",
+    "education_level",
+    "employee_id",
+    "end_date",
+    "first_name",
+    "full_name",
+    "gender",
+    "hire_date",
+    "last_name",
+    "management_role",
+    "marital_status",
+    "position_id",
+    "position_title",
+    "salary",
+    "store_id",
+    "supervisor_id"
+  ],
+  salary: [
+    "currency_id",
+    "department_id",
+    "employee_id",
+    "overtime_paid",
+    "pay_date",
+    "salary_paid",
+    "vacation_accrued",
+    "vacation_used"
+  ]
+};
+
+const getTable = (id: string) => ({
+  id
+});
+
+const getColumns = (tableId: string) => {
+  const columns = columnsMapping[tableId] || [];
+  return columns.map(columnId => ({
+    id: columnId,
+    tableId
+  }));
+};
+
+const metadataProvider = {
+  getTable: getTable,
+  getColumns: getColumns
+};
+
+// function to log lineage when testing
+// eslint-disable-next-line @typescript-eslint/no-unused-vars
+function logLineage<TableData extends { id: string }, ColumnData extends { id: string }>(
+  lineage: Lineage<TableData, ColumnData>
+) {
+  console.log(JSON.stringify(lineage, undefined, 2));
+}
+
+describe("antlr", () => {
+  it("should build parse tree", () => {
+    const input = "select * from emp";
+    const inputStream = new UppercaseCharStream(CharStreams.fromString(input));
+    const lexer = new SqlBaseLexer(inputStream);
+    const tokens = new CommonTokenStream(lexer);
+    const parser = new SqlBaseParser(tokens);
+    parser.buildParseTree = true;
+    const tree = parser.statement();
+    expect(tree.toStringTree()).toBe(
+      "(select*fromemp (select*fromemp (select*fromemp (select*fromemp (select*fromemp (select* select (* (* (* (* (* (* *))))))) (fromemp from (emp (emp (emp (emp (emp (emp emp)) )) )))))) ))"
+    );
+  });
+
+  it("should work with aliases", () => {
+    const sql =
+      "select sq.account_id as account_id, e.employee_id as employee_id \n" +
+      "from (select account_id as account_id, account_type as account_type from account as d) as sq, \n" +
+      "   employee as e";
+    const lin = getLineage(sql, metadataProvider);
+    expect(lin.filter(e => e.type == "edge")).toHaveLength(4);
```

Comment:
I've chosen to check number of edges because it means that columns are found and really matching. And you can count the number of edges from the original query by hand. On the other hand it doesn't assume any order of the edges and relations, as well as relation and edge Ids which can change. I think that's good balance between actually checking correctness and test robustness and simplicity. What's you proposal how it could look like?

Reply:
My proposal is to simply test the output (like here https://github.com/contiamo/rhombic/blob/master/src/index.getLineage.test.ts#L67-L173), my main concern here is that I'm quite sure I can play with your implementation without breaking the tests (so if somebody miss a refactor or something in a future feature, I have no guarantee than the actual behaviours are still working…), the idea is to pick some interesting examples and make sure they are working as expected (with the correct relations and metadata). Checking the number of edges can be a good first test but this is definitely not enough to ensure everything is working as we expect.